src/lib/accessibility/FocusHandling/FocusableItem.ts

```diff
@@ -0,0 +1,113 @@
+export interface IFocusableItemProps {
+  isFocused: boolean
+
+  isFirstElement: boolean
+  isLastElement: boolean
+
+  onMovePrevious: () => void
+  onMoveNext: () => void
+  onMoveFirst: () => void
+  onMoveLast: () => void
+  onEnter: () => void
+  onSpace: () => void
+  onEsc: () => void
+}
+
+export interface IFocusableItemState {
+  shouldSubContainerBeOpened: boolean
+  isLastOpened: boolean
+}
+
+export type SetStateDelegate<P, S> = <K extends keyof S>(
+  state:
+    | ((prevState: Readonly<S>, props: Readonly<P>) => Pick<S, K> | S | null)
+    | (Pick<S, K> | S | null),
+  callback?: () => void,
+) => void
+
+export class FocusableItem<P extends IFocusableItemProps, S extends IFocusableItemState> {
+  constructor(
+    private getProps: () => P,
+    private setState: SetStateDelegate<P, S>,
+    private initState: (state: IFocusableItemState) => void,
+  ) {
+    this.initState({
+      shouldSubContainerBeOpened: false,
+      isLastOpened: false,
+    })
+  }
+
+  public focus(focusableElement: HTMLElement) {
+    if (this.getProps().isFocused) {
+      focusableElement.focus()
+    }
+  }
+
+  public movePrevious() {
+    if (this.getProps().isFirstElement || !this.getProps().isFocused) {
+      return
+    }
+
+    this.getProps().onMovePrevious()
+  }
+
+  public moveNext() {
+    if (this.getProps().isLastElement || !this.getProps().isFocused) {
+      return
+    }
+
+    this.getProps().onMoveNext()
+  }
+
+  public moveFirst() {
+    if (this.getProps().isFirstElement || !this.getProps().isFocused) {
+      return
+    }
+
+    this.getProps().onMoveFirst()
+  }
+
+  public moveLast() {
+    if (this.getProps().isLastElement || !this.getProps().isFocused) {
+      return
+    }
+
+    this.getProps().onMoveLast()
+  }
+
+  public enter() {
+    this.setState({ isLastOpened: false })
+
+    if (!this.getProps().isFocused) {
+      return
+    }
+
+    this.setState({
+      shouldSubContainerBeOpened: true,
+      isLastOpened: true,
```

Comment:
Also as these are not used anywhere at the moment, it is difficult for me to understand their purpose.

Reply:
good point, Removed this part for now, until we have the real example to use it